Smooth Movement 5.2: God's Dumbest Code Edition (PLATNIUM) (BESTSELLING)

LemonInTheDark opened the pull request:
## About The Pull Request

I HAVE RETURNED! #89025 made manifest!

Hey! I'm back on my bullshit.
This pr was born out of the nausea reported during testing of #85491. In the process of working out why gliding stuttered so much I discovered 2 major issues with our movement visuals. I'll describe them in order, alongside their solutions, and then get into the fun bonuses here.

### Hey wait a minute this sucks

Ok so like, base principles, the idea of gliding is you tell the client "hey this mob is moving, animate it moving from A to B at X pixels per tick".  This is fine, except we don't want mobs to stop and start when moving in game, and if the game lags we will end up delaying their next move by X ms.

Easy solution then, track the average lag and multiply glide rates to correct them.
This is fine, and about what I'm doing, it's just our old method of getting "average lag" was to, once every 10 seconds, get the ratio between real and game time over the PRECEDING 10 seconds. 

Two problems here:
1. We're working off REALLY outdated information, so we're likely to finish early and just sit on the tile for a few game ticks, or worse jump forward because we're moving too slow. This feels jerky and happens CONSTANTLY.
2. Because this is essentially a raw average, we don't care if the lag is sustained (the thing that matters) or due to 1/2 lag spikes (say, hard deletes). This, like the above, would cause us to underestimate how fast it takes for the average next game tick to come out, leading to jumping. This also happens all the time and feels horrid.

Ok so we need better numbers.
Byond provides a measure of the % of the tick the last gametick needed to finish, which would be perfect for this. It's unfortunately averaged in a constant size 16 moving average, and we need the raw value. I've tried unrolling this manually but it persists across restarts and drops data, so that's impossible. Absolucy saved my ass and did some [memory hacking](https://github.com/Absolucy/auxcpu) to get the real numbers for me (Insanely based). This does make us dependent on auxtools, but it's minimal and IN THEORY lumlum will expose this someday too (Bump my [feature request](https://www.byond.com/forum/post/2968149) please)

We can then track cpu usage over the past X ticks, and remove outliers and build an average we can use to multiply against new glides. I am decently happy with this though I could use some pointers on the averaging math.

Of note: I tried glide correction, but the jerking and slowdown felt worse/the same as the impacts of lag spikes normally, and it led to extra complexity/cost. I am not 100% confident in my implementation so perhaps there's stuff to do here.

This all works pretty damn good, but it doesn't entirely solve stuttering, even in cases without overtime/client fps drops.

### Oh my god BYOND is stupid

I noticed (alongside SmArtKar) that the game is FAR smoother before roundstart then after, even if we aren't going into overtime. I was curious about why, and after MUCH digging around, realized that subsystems consuming cpu caused the stuttering.

This was fucking weird. looked more into it, the stutters didn't happen at the same time for all mobs, though they did seem to be roughly consistent across clients maybe? Did a bunch of frame by frame work, died due to my burning hate for this community, came back to life, did even more frame by frame work, and discovered that what was going on was VARIATION in cpu usage dropping frames.

This isn't an us issue, it's an [ENGINE issue](https://www.byond.com/forum/post/2959355). It occurs locally and remotely, on ss13 code and on raw byond. 

What's going on is (to my understanding didn't fully get lummy's explanation) if the time BETWEEN SendMaps() calls is too different, we may end up skipping client frames ([Background Reading](https://github.com/tgstation/tgstation/blob/master/.github/guides/TICK_ORDER.md)). See the image below.

![art](https://github.com/user-attachments/assets/ff087463-0d3d-4a67-bbea-de04313fe471)

Lummy seemed to be unsure of if he was going to be able to fix this (something about glide compensating) so the only option left to us outside of Deal With ItTM is to ensure the gap between visual data getting sent is as consistent as possible.

This means we need to consume up to the highest cpu values that we SAFELY can each tick. This is dangerous because we don't have a hookpoint after verb/map cpu work so we can't really avoid eating into their space. I've setup code to target a usage level, and then account for passive cost ABOVE that threshold (to hopefully avoid overtime). I'm currently pinning about 90%, so there should be some room for spikes in here too. Need to test on live to know for sure though.

In short we're essentially burning cpu time on literally nothing to reduce visual lag. I hate it here.

Here's a demonstration of the behavior, to prove my bullshit

https://file.house/AQDCF0YHk7YU83I2kH_-Tw==.mp4

### Debug Tooling

You'll notice a bunch of shit on the top left in the above video, as a part of working this shit out I also added some debug tools.
There's a new verb that enables cpu debug maptext, which displays the tick number, correction info, total cpu, pre Tick(), Tick(), ~maptick, and ~verbs (both live, minimum and maximum over a window) (also the drift between reported cpu and the average, for safety in testing).

In addition I've added some global values that allow for modifying cpu floor/ceiling and spiking for testing lag (also modifying the correction values).

Finally, a large part of the code diffs in this pr is support for a graphing system, which I have plugged CPU, pre Tick() CPU and Maptick+Verbs into. This allows us to in real time monitor cpu usage and gain some understanding of what is actually going on.

Here's a video of said graph showing how to use it/what it can do.

https://file.house/0dtAtol5bTdOTWi80Z535w==.mp4

### Update: I have crashed in. 

[Implements verb tracking system, shims all verbs/verblikes I can](https://github.com/tgstation/tgstation/pull/91052/commits/25a0a019262a82c6d314a168fd1c0bb69115d8e9)
[25a0a01](https://github.com/tgstation/tgstation/pull/91052/commits/25a0a019262a82c6d314a168fd1c0bb69115d8e9)

Yea ok I maybe went a bit overboard here.

BYOND does not actually include verb cost in world.cpu, or any other metric for that matter.
This is because verbs do not get worked on as a lump. network events can happen ANYTIME and are processed as soon as the thread gets to them.

This means we can have large gaps between verbs being invoked, which would make any report of their cost unclear.

Sooooooo I'm shimming them, tracking their start/stop time and am going to feed this into my graphing software to gain a full understanding of where our cpu time is going.

This is mildly messy, mostly because I need to account for different ways of handling sleep depending on how a verb is invoked. I'm sorry.

[Drastically expands existing CPU visualization](https://github.com/tgstation/tgstation/pull/91052/commits/60f98f0b9d88dba32571acd9fa5a1ac46b487651)
I went a bit insane.

We now track and display:
Early sleeping cost
Master controller cost (hovering bars displays subsystem breakdown)
Late sleeping cost
Tick() cost
Maptick cost
Normal CPU (world.cpu values)
Verb cost
Verb TIMINGS (IE: what ran when, hovering displays the ran verbs and their costs)
A complete view of the tick, with hover breakdowns for each component)

Refactored tooltip code a bit to make this work.
Also, added running averaging for verb cost, as a global list. We should be able to use SOMETHING like this to autoqueue expensive verbs.

Here's an updated video with the new toys

https://file.house/mqnIhq7spUZAOYbX_lZxFw==.mp4

### Why Is This Drafted?

I need to run it on live to make sure everything works out ok.

## Why It's Good For The Game

The game should be smooth. Duh. I can't play/code anymore because the jerkyness makes me depressed. 

Also, we should have more monitoring tools for cpu usage/overtime so we can actually understand our tick allocation and solve problems on live.

## Changelog
:cl: LemonInTheDark, Absolucy
fix: The game should be significantly smoother, both movement gliding and general consistent frame times
admin: Added CPU debugging tools, including.... 10? different types of graphs. See "Toggle Cpu Controls"
/:cl:

LemonInTheDark committed:
Implements better (maybe) lag compensation for gliding

Our existing lag compensation is utter dogwater, it works off a 10
second average that was taken between 0 and 10 seconds ago. Overshooting
leads to stutter steps, which are highly unpleasent, and easy to achive
since any spike in overtime will lead to an overshoot.

What I'm doing here is injesting cpu_usage, deaveraging it using a
world/Tick override (it has an 16 frame moving avg), then taking an
average of that and dropping noisy bits.

I am not the best at stats so this implementation might kinda suck, but
it's better then what we have now.

I'm also trying out spike correction, IE accounting for the delay a
lagspike causes.

This is less of an issue then jumping is because you just end up staying
on a turf for a few ticks longer, but if we cna just slow people down
some I think it's worthwhile.

I am less confident in this concept/implementation then I am with the
tick correction, but we'll see.

I'm putting this up to test on live/motion sick people, should not be
merged.

LemonInTheDark committed:
what the fuck

LemonInTheDark committed:
size increase

LemonInTheDark committed:
Changes avg method to keep a fix sized window

I hate byond

LemonInTheDark committed:
whoop

LemonInTheDark committed:
IT WORKED!!!

LemonInTheDark committed:
oh whoops\

LemonInTheDark committed:
Lemon's amazing magic mob mover

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into dynamic-gliding

LemonInTheDark committed:
Implements potentially better correction and debugging

Glide corrects as if we are a tick in the past (since our info is a tick
out of date), draws more debug info (and draws it to a hidden plane
master) and locks active correction behind a global value

LemonInTheDark committed:
Better debug + glide text

LemonInTheDark committed:
Fixes people seeing my shit, better control, more complete tick readout to shittest

LemonInTheDark committed:
fuck

LemonInTheDark committed:
Noise control, verb

Floors real cpu values with the tick usage/map cpu from the last tick
This serves as a worst case minimum and should bound any error
introduced by floating point bullshit to a particular bound (since error
compounds here)

Adds a verb that toggles on/off the cpu display, so players are not
subjected to it for no reason

LemonInTheDark committed:
Reserves more time for verbs, adds more tracking, autocorrection mode

Verbs take an undefined amount of time (we should really track/manage
this better) and if we're gonna be consuming 100% of the tick we should
limit that (more on that later)

We have map cpu, tick cost, and deaveraged cpu, so we can work out
ROUGHLY how much time we spend on verbs + other bullshit. Should display
that

Adds a new global that enables cpu stabailization, queuing up
/world/Tick() to consume a constant amount of cpu (since that stabalizes
the deltas between sendmaps calls, reducing signficantly visible lag)

LemonInTheDark committed:
Rips out all the testing stuff, proper plane master support

LemonInTheDark committed:
Reorgs somewhat

LemonInTheDark committed:
whoop

LemonInTheDark committed:
whoop2

LemonInTheDark committed:
missed a spot

LemonInTheDark committed:
fuck

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into the-highway-or-the-hangman

LemonInTheDark committed:
pain

LemonInTheDark committed:
yeet

LemonInTheDark committed:
lint?

LemonInTheDark committed:
whoop

LemonInTheDark committed:
whoop

LemonInTheDark committed:
whoop

LemonInTheDark committed:
New form of error correction using the font of truth that is world.cpu

Instead of flooring the cpu values (which allowed spikes to occasionally
just sit undisterbed forever) we correct them as we go using an expanded
world.cpu value. this will not be precise but if each value as we
iterate is corrected then the problem child will always JUST be the new
reading, should constrain error effectively

LemonInTheDark committed:
adds control over glide type, more debug shit, canery for negatives

LemonInTheDark committed:
bump er up

LemonInTheDark committed:
Updates to a wrapper datum, partially impls absolucy's auxtools thing (I love you)

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into maybe-god-isnt-the-right-word-but-i-believe-in-you

LemonInTheDark committed:
Adds Graphing API + Full Lucy Impl

I've built out some code to draw graphs off game data, it's maybe a bit
oversubtyped but I want something I can pull out of in future.
I'm trying to be sort of modular here, so it's easy to reuse/expand on
in future. It's not perfect but it'll do for this.

Have removed some now irrelevant verb logic, added new controls for the
graph.

I'm keeping some of the tick by tick logic in just for stats, they're
not needed for behavior really.

LemonInTheDark committed:
Cleanup, Autocorrection Support, Multiple Displays

LemonInTheDark committed:
rename Tick()

LemonInTheDark committed:
bit more readable

LemonInTheDark committed:
commits dll

LemonInTheDark committed:
whoops missed it

Kapu1178 commented:
you're ugly you're stupid im gonna kill you, now give me 200 dollars

LemonInTheDark commented on code/game/world.dm:
```diff
@@ -147,6 +151,347 @@ GLOBAL_VAR(restart_counter)
+	var/last_index = tick_info.cpu_index
+	var/full_time = TICKS2DS(INTERNAL_CPU_SIZE) / 10 // convert from ticks to seconds
+
+	maptext = "<div style=\"background-color:#FFFFFF; color:#000000;\">\
```
we ball actually

LemonInTheDark committed:
autoenable features on join, disables autotoggle (kills dub)

LemonInTheDark committed:
more trace points

flleeppyy commented:
everyone can learn a thing or two from these amazing commit messages

LemonInTheDark committed:
into the void

LemonInTheDark committed:
shift down

LemonInTheDark committed:
wrong name

Wallemations commented:
> Copilot reviewed 19 out of 19 changed files in this pull request and generated 2 comments.

When the hell did copilot learn to understand .dm files

LemonInTheDark committed:
ckey bypass

LemonInTheDark committed:
whoops the culling should actually like, work I think (personally)

LemonInTheDark committed:
adds a line displaying the MC's target cpu (to avoid overtime)

LemonInTheDark committed:
Implements test replacing MC threshold with calculated consumption threshold

Cleans up the code further, adds a line displaying the consumption
limit, uses basically a max algorithem for cpu thresholding to avoid
spikes pushing above our limits

LemonInTheDark committed:
mc data should go in properly

LemonInTheDark committed:
makes the longer window sizes actually work

LemonInTheDark committed:
fuckfuckfuckfuck

LemonInTheDark committed:
bigger fuck

LemonInTheDark committed:
code should actually work, also, adds unused prime counting code

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into maybe-god-isnt-the-right-word-but-i-believe-in-you

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into maybe-god-isnt-the-right-word-but-i-believe-in-you

LemonInTheDark committed:
Implements verb tracking system, shims all verbs/verblikes I can

Yea ok I maybe went a bit overboard here.

BYOND does not actually include verb cost in world.cpu, or any other
metric for that matter.
This is because verbs do not get worked on as a lump. network events can
happen ANYTIME and are processed as soon as the thread gets to them.

This means we can have large gaps between verbs being invoked, which
would make any report of their cost unclear.

Sooooooo I'm shimming them, tracking their start/stop time and am going
to feed this into my graphing software to gain a full understanding of
where our cpu time is going.

This is mildly messy, mostly becuase I need to account for different
ways of handling sleep depending on how a verb is invoked. I'm sorry

LemonInTheDark committed:
Drastically expands existing CPU visualization

I went a bit insane.

We now track and display:
Early sleeping cost
Master controller cost (hovering bars displays subsystem breakdown)
Late sleeping cost
Tick() cost
Maptick cost
Normal CPU (world.cpu values)
Verb cost
Verb TIMINGS (IE: what ran when, hovering displays the ran verbs and
their costs)
A complete view of the tick, with hover breakdowns for each component)

Refactored tooltip code a bit to make this work.
Also, added running averaging for verb cost, as a global list. We should
be able to use SOMETHING like this to autoqueue expensive verbs.

LemonInTheDark committed:
% output not ms

SmArtKar commented:
any shot at this going through?

LemonInTheDark commented:
I am doing that thing I do where all my prs now balloon into complexity hell as a joke.
This basically works fine, I want to implement autoverb queuing before merging this since we're dancing with death a lot more often now. there's also this https://gist.github.com/LemonInTheDark/028d12dc370acded09af0e1405f23afa but most of that should happen at a time that is not now in a place that is not this pr

LemonInTheDark commented:
I should be able to finish this in the next few weeks, just kinda gettin rolled rn
